docs/RFCS/sql_batching.md

```diff
@@ -0,0 +1,501 @@
+we use for constructing the dependency graph in our `CommandQueue`.
+
+Again, we can extend this reasoning to groups of SQL statements. We also say that a
+group of statements has a **read-set (R)** and **write-set (W)** which are respectively
+the unions of the read-set and the write-set of all statements in the group. When
+determining if a group of statements is independent of another group of statements,
+we can use the same rules as above using each entity's read and write-sets.
+
+It is easy to reason that SELECT statements and other SQL statements that are strictly
+read-only are always independent of one another, regardless of any overlap in the data
+they touch. These statements perform no externally-visible state mutation, and will not
+interfere with one another (from a correctness standpoint at least, see
+[below](#disrupting-data-locality-and-other-unintended-performance-consequences) for a
+discussion on indirect statement interaction). Using our previous notation, each statement
+will have an empty write-set (W<sub>1</sub> = W<sub>2</sub> = Ø), so all both intersections
+in the statement independence rules will produce the empty set.
+
+Perhaps more interestingly, we can also reason that two statements that perform mutations
+(INSERT, UPDATE, DELETE, etc.) but that touch disjoint sets of data are also independent
+of one another. It is important to note here that the set of data that a statement _touches_
+includes any that the statement mutates and any that the statement reads while performing
+validation and filtering.
+
+Taking this a step further, we can use our statement independence rules to find that two
+mutating statements can be independent even if their touched data sets overlap, as long
+as only their read-sets overlap. A practical example of this result is two statements that
+insert rows into two different tables, but each validates a foreign key in the same third
+table. These two statements are independent because only their read-sets overlap.
+
+Finally, we can reason that if there is overlap between the touched sets of data between
+any two statements where at least one of the two performs mutation in the overlapping region,
+then the statements are not independent.
+
+Examples:
+
+```
+SELECT * FROM a
+SELECT * FROM b
+SELECT * FROM a
+```
+All three statements are **independent** because they have empty write-sets.
+
+```
+INSERT INTO a VALUES (1)
+INSERT INTO b VALUES (1)
+```
+Statements are **independent** because they have disjoint write-sets and empty read-sets.
+
+```
+UPDATE a SET x = 1 WHERE y = false
+UPDATE a SET x = 2 WHERE y = true
+```
+Statements are **independent** because they have disjoint write-sets and their read-sets
+do not overlap with the other's write-sets. (note that parallelizing this example will not
+actually be supported by the initial proposal, see [Dependency Granularity](#dependency-granularity))
```

Comment:
`UPSERT` feels similar to `INSERT` in that it should be possible to look at the `VALUES` to determine if there are dependencies. And given that `UPSERT` is short-hand for `UPDATE`, I wonder if we could look for the simple single-row-`WHERE`-matches-primary-key pattern. I bring this up because the motivating example has 2 UPDATEs to a single table to different rows.